pysnark/runtime.py

```diff
@@ -204,97 +204,203 @@ def __ge__(self, other): return (self-other).check_positive()
+            if other.value == 0:
+                raise ValueError("Division by zero")
+            elif is_guard() and (self.value % other.value == 0):
+                res = PrivVal(self.value // other.value)
             elif ignore_errors():
                 res = PrivVal(0)
             else:
                 raise ValueError(str(self.value) + " is not properly divisible by " + str(other.value))
             add_constraint(other, res, self)
             return res
-        elif is_base_value(other):
-            if self.value % other != 0:
-                raise ValueError(str(self.value) + " is not properly divisible by " + str(other))
-            return LinComb(self.value/other, self.lc*backend.fieldinverse(other))
-        else:
-            return NotImplemented
+
+        return NotImplemented
          
     def __floordiv__(self, other):
-        """ Division with rounding """
-        return self.__divmod__(other)[0]
+        """
+        Divides a LinComb with an integer or another LinComb using floor division
+        Costs 0 constraints to divide with an integer
+        Costs 1 constraint to divide with a LinComb
+        """
+        res = self.__divmod__(other)
+        if res is NotImplemented:
+            return NotImplemented
+        return res[0]
 
     def __mod__(self, other):
-        if other&(other-1)==0:
-            # this is faster for powers of two
-            return LinComb.from_bits(self.to_bits()[:other.bit_length()-1])
-        
-        return self.__divmod__(other)[1]
+        """
+        Returns the remainder of a LinComb divided with an integer or another LinComb
+        Costs 0 constraints to divide modulo an integer
+        Costs 1 constraint to divide modulo a LinComb
+        """
+        res = self.__divmod__(other)
+        if res is NotImplemented:
+            return NotImplemented
+        return res[1]
         
     def __divmod__(self, divisor):
-        """ Division by public value """
-        
-        if not is_base_value(divisor): return NotImplemented
-        
-        if divisor==0: raise ValueError("division by zero")
-        
-        quo = PrivVal(self.value//divisor)
-        rem = PrivVal(self.value-quo.value*divisor)
- 
-        rem.assert_positive(divisor.bit_length())
-        if divisor&(divisor-1)!=0: rem.assert_lt(divisor) # not needed for powers of two
-        quo.assert_positive()
-        (self-divisor*quo-rem).assert_zero()
-        
-        return (quo,rem)
+        """
+        Divides a LinComb with an integer or another LinComb and returns the quotient and the remainder
+        Costs 0 constraints to divide with an integer
+        Costs 3 constraints to divide with a LinComb
+        """
+        if isinstance(divisor, int):
+            if divisor == 0:
+                raise ValueError("Division by zero")
+            quo = ConstVal(self.value // divisor)
+            rem = ConstVal(self.value % divisor)
+            return (quo, rem)
```

Comment:
Here it seems that there is no check that the computed quo and rem are correct (assert_positives and (self-divisor*quo-rem).assert_zero(). Also, shouldn't quo and rem be PrivVals (witness values) instead of ConstVals (publicly known constants)?